app/src/main/java/io/apicurio/registry/storage/importing/v2/SqlDataUpgrader.java

```diff
@@ -0,0 +1,270 @@
+
+    protected RegistryStorageContentUtils utils;
+
+    protected final RegistryStorage storage;
+
+    protected final boolean preserveGlobalId;
+
+    protected final boolean preserveContentId;
+
+    // To handle the case where we are trying to import a version before its content has been imported
+    protected final List<ArtifactVersionEntity> waitingForContent = new ArrayList<>();
+
+    // To handle the case where we are trying to import a comment before its version has been imported
+    private final List<CommentEntity> waitingForVersion = new ArrayList<>();
+
+    // ID remapping
+    protected final Map<Long, Long> globalIdMapping = new HashMap<>();
+    protected final Map<Long, Long> contentIdMapping = new HashMap<>();
+
+    // To keep track of which versions have been imported
+    private final Set<GAV> gavDone = new HashSet<>();
+
+    public SqlDataUpgrader(Logger logger, RegistryStorageContentUtils utils, RegistryStorage storage,
+            boolean preserveGlobalId, boolean preserveContentId) {
+        super(logger);
+        this.utils = utils;
+        this.storage = storage;
+        this.preserveGlobalId = preserveGlobalId;
+        this.preserveContentId = preserveContentId;
+    }
+
+    @Override
+    protected void importArtifactRule(ArtifactRuleEntity entity) {
+        try {
+            io.apicurio.registry.utils.impexp.v3.ArtifactRuleEntity newEntity = io.apicurio.registry.utils.impexp.v3.ArtifactRuleEntity
+                    .builder().type(entity.type).artifactId(entity.artifactId)
+                    .configuration(entity.configuration).groupId(entity.groupId).build();
+            storage.importArtifactRule(newEntity);
+            log.debug("Artifact rule imported successfully: {}", entity);
+        } catch (Exception ex) {
+            log.warn("Failed to import artifact rule {}: {}", entity, ex.getMessage());
+        }
+    }
+
+    @Override
+    public void importArtifactVersion(ArtifactVersionEntity entity) {
+        try {
+            // Content needs to be imported before artifact version
+            if (!contentIdMapping.containsKey(entity.contentId)) {
+                // Add to the queue waiting for content imported
+                waitingForContent.add(entity);
+                return;
+            }
+
+            entity.contentId = contentIdMapping.get(entity.contentId);
+
+            var oldGlobalId = entity.globalId;
+            if (!preserveGlobalId) {
+                entity.globalId = storage.nextGlobalId();
+            }
+
+            io.apicurio.registry.utils.impexp.v3.ArtifactVersionEntity newEntity = io.apicurio.registry.utils.impexp.v3.ArtifactVersionEntity
+                    .builder().createdOn(entity.createdOn).description(entity.description)
+                    .labels(entity.labels != null
+                        ? entity.labels.stream().collect(Collectors.toMap(label -> label, label -> label))
+                        : Collections.emptyMap())
```

Comment:
V2 versions can have both labels and properties.  I think properties should be brought over as-is, and labels should be added with `null` values.

Reply:
OK, I wasn't sure about the value, so I decided to use the label value from v2.